drv/meanwell/src/main.rs

```diff
@@ -0,0 +1,179 @@
+//!
+#![no_std]
+#![no_main]
+
+use drv_meanwell_api::MeanwellError;
+use idol_runtime::NotificationHandler;
+use idol_runtime::RequestError;
+use userlib::*;
+
+use drv_stm32xx_sys_api as sys_api;
+
+struct ServerImpl {
+    deadline: u64,
+    led_on: bool,
+}
+
+cfg_if::cfg_if! {
+    if #[cfg(any(target_board = "gimletlet-1", target_board = "gimletlet-2"))] {
+        const MEANWELL_PINS: &[sys_api::PinSet] = &[
+            sys_api::Port::B.pin(14),
+            sys_api::Port::B.pin(15),
+            sys_api::Port::D.pin(8),
+            sys_api::Port::D.pin(9),
+            sys_api::Port::D.pin(10),
+            sys_api::Port::D.pin(11),
+            sys_api::Port::D.pin(12),
+        ];
+
+        // In keeping with the Meanwell's blue indicator LED, we use only the
+        // blue LED to indicate that this is a Meanwell management Gimletlet.
+        const LED_INDEX: usize = 3;
+    } else {
+        compile_error!("unsupported Meanwell board");
+    }
+}
+
+fn set(index: usize, val: bool) -> Result<(), RequestError<MeanwellError>> {
+    use drv_stm32xx_sys_api::*;
+
+    let sys = SYS.get_task_id();
+    let sys = Sys::from(sys);
+
+    if index >= MEANWELL_PINS.len() {
+        Err(MeanwellError::NotPresent.into())
+    } else {
+        match sys.gpio_set_to(MEANWELL_PINS[index], val) {
+            Ok(_) => Ok(()),
+            Err(_) => Err(MeanwellError::GpioError.into()),
+        }
+    }
+}
+
+fn get(index: usize) -> Result<bool, RequestError<MeanwellError>> {
+    use drv_stm32xx_sys_api::*;
+
+    let sys = SYS.get_task_id();
+    let sys = Sys::from(sys);
+
+    if index >= MEANWELL_PINS.len() {
+        Err(MeanwellError::NotPresent.into())
+    } else {
+        match sys.gpio_read(MEANWELL_PINS[index]) {
+            Ok(val) => Ok(val != 0),
+            Err(_) => Err(MeanwellError::GpioError.into()),
+        }
+    }
+}
+
+impl idl::InOrderMeanwellImpl for ServerImpl {
+    fn power_on(
+        &mut self,
+        _: &RecvMessage,
+        index: usize,
+    ) -> Result<(), RequestError<MeanwellError>> {
+        set(index, true)
+    }
+
+    fn power_off(
+        &mut self,
+        _: &RecvMessage,
+        index: usize,
+    ) -> Result<(), RequestError<MeanwellError>> {
+        set(index, false)
+    }
+
+    fn is_on(
+        &mut self,
+        _: &RecvMessage,
+        index: usize,
+    ) -> Result<bool, RequestError<MeanwellError>> {
+        get(index)
+    }
+}
+
+task_slot!(USER_LEDS, user_leds);
+task_slot!(SYS, sys);
+
+const TIMER_MASK: u32 = 1 << 0;
+const TIMER_INTERVAL_LONG: u64 = 900;
+const TIMER_INTERVAL_SHORT: u64 = 100;
+
+impl NotificationHandler for ServerImpl {
+    fn current_notification_mask(&self) -> u32 {
+        TIMER_MASK
+    }
+
+    fn handle_notification(&mut self, _bits: u32) {
+        let user_leds =
+            drv_user_leds_api::UserLeds::from(USER_LEDS.get_task_id());
+
+        if !self.led_on {
+            user_leds.led_on(LED_INDEX).unwrap();
```

Comment:
Nit: `UserLeds` exposes an `led_toggle` command

Reply:
If you don't mind, I think I want to leave this as an explicit on to assure that its state is what the driver thinks it should be. (The LED is on for 100 ms and off for 900 ms.)